homeassistant/components/sunsynkweb/sensor.py

```diff
@@ -0,0 +1,183 @@
+"""Sensors for the sunsynk web api."""
+
+import decimal
+import logging
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import PERCENTAGE, UnitOfEnergy, UnitOfPower
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from . import DOMAIN
+from .coordinator import PlantUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SunSynkApiSensor(CoordinatorEntity, SensorEntity):
+    """Parent class for sunsynk api exposing sensors.
+
+    The sensors expose the sum of the power across all inverters
+    for plants that have more than one inverter.
+    State of charge is normally shared between inverters, so this will take
+    the maximum of the state of charge.
+    """
+
+    def __init__(self, agg_func, coordinator, description) -> None:
+        """Initialise the common elements for sunsynk web api sensors."""
+        CoordinatorEntity.__init__(self, coordinator, context=coordinator)
+        self.coordinator: PlantUpdateCoordinator = coordinator
+        self.entity_id = f"sensor.sunsynk_{description.key}_{sum(p.id for p in coordinator.cache.plants)}"
+        self.name = description.name
+        self.entity_description = description
+        self.agg_func = agg_func
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        if self.coordinator.cache is not None:
+            self.native_value = self.agg_func(
+                decimal.Decimal(getattr(i, self.entity_description.key))
+                for i in self.coordinator.cache.plants
+            )
+        else:
+            _LOGGER.debug("Not updating sensor as coordinator cache is empty. ")
+        self.async_write_ha_state()
+
+
+SENSOR_DESCRIPTIONS = [
+    (
+        sum,
+        SensorEntityDescription(  # ""A gauge for battery power."""
+            device_class=SensorDeviceClass.POWER,
+            native_unit_of_measurement=UnitOfPower.WATT,
+            key="battery_power",
+            name="Solar supply battery power (instantaneous)",
```

Comment:
you could extend the `SensorEntityDescription` (by building your own dataclass, look at e.g. `lamarzocco` for examples) to include a `agg_function` key, making it a bit easier to read imo

Reply:
updated